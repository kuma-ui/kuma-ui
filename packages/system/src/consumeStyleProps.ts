import { AnimationProps, animationMappings } from "./props/animation";
import { SpaceProps, spaceMappings } from "./props/space";
import { TypographyProps, typographyMappings } from "./props/typography";
import { LayoutProps, layoutMappings } from "./props/layout";
import { ColorProps, colorMappings } from "./props/color";
import { FlexProps, flexMappings } from "./props/flex";
import { BorderProps, borderMappings } from "./props/border";
import { OutlineProps, outlineMappings } from "./props/outline";
import { PositionProps, positionMappings } from "./props/position";
import { ShadowProps, shadowMappings } from "./props/shadow";
import { ThemeSystemType, ResponsiveStyle } from "./types";
import { styleCache } from "@kuma-ui/sheet";
import { GridProps, gridMappings } from "./props/grid";
import { ListProps, listMappings } from "./props/list";
import { EffectProps, effectMappings } from "./props/effect";
import { TextProps, textMappings } from "./props/text";
import { FontProps, fontMappings } from "./props/font";
import { MaskProps, maskMappings } from "./props/mask";
import { ColumnProps, columnMappings } from "./props/column";
import { BackgroundProps, backgroundMappings } from "./props/background";
import { FilterProps, filterMappings } from "./props/filter";
import { StyledKeyType } from "./keys";
import { applyResponsiveStyles } from "./responsive";
import { toCssUnit } from "./toCSS";

export type StyledProps<T extends ThemeSystemType = ThemeSystemType> =
  TypographyProps<T> &
    FontProps<T> &
    ColorProps<T> &
    SpaceProps<T> &
    AnimationProps &
    TextProps &
    LayoutProps<T> &
    FlexProps<T> &
    BorderProps<T> &
    OutlineProps &
    PositionProps &
    ShadowProps &
    GridProps<T> &
    ListProps &
    EffectProps &
    MaskProps &
    ColumnProps<T> &
    BackgroundProps &
    FilterProps;

export type StyleFunction = (props: StyledProps) => ResponsiveStyle;

// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment -- FIXME
const styleMappings: Record<StyledKeyType, string> = Object.assign(
  {},
  animationMappings,
  spaceMappings,
  typographyMappings,
  layoutMappings,
  colorMappings,
  flexMappings,
  borderMappings,
  outlineMappings,
  positionMappings,
  shadowMappings,
  gridMappings,
  listMappings,
  effectMappings,
  textMappings,
  fontMappings,
  maskMappings,
  columnMappings,
  backgroundMappings,
  filterMappings,
);

export const consumeStyleProps = (props: StyledProps): ResponsiveStyle => {
  const cacheKey = JSON.stringify(props);

  const cachedStyles = styleCache.get(cacheKey);
  if (cachedStyles) {
    return cachedStyles;
  }

  let base = "";
  const media: { [breakpoint: string]: string } = {};
  for (const key in props) {
    const cssValue = props[key as StyledKeyType];
    if (cssValue == null) continue;

    const properties = styleMappings[key as StyledKeyType]?.split(",") ?? [];
    for (const property of properties) {
      const responsiveStyles = applyResponsiveStyles(
        property,
        cssValue,
        (value) => toCssUnit(key, value),
      );
      base += responsiveStyles.base;
      for (const [breakpoint, css] of Object.entries(responsiveStyles.media)) {
        if (media[breakpoint]) {
          media[breakpoint] += css;
        } else {
          media[breakpoint] = css;
        }
      }
    }
  }

  const combinedStyles = { base, media };

  styleCache.set(cacheKey, combinedStyles);

  return combinedStyles;
};
