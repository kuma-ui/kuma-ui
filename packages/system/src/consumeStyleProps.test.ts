import { consumeStyleProps, StyledProps } from "./consumeStyleProps";
import { describe, expect, test } from "vitest";

function consumeAndTest(props: StyledProps) {
  // Act
  const styles = consumeStyleProps(props);
  // Assert
  expect(styles).toMatchSnapshot();
}

describe("consumeStyleProps function", () => {
  test("should consume animation props", () => {
    consumeAndTest({
      animation: "3s ease-in 1s infinite reverse both running slidein",
      animationComposition: "add",
      animationDelay: "1s",
      animationDirection: "reverse",
      animationDuration: "1s",
      animationFillMode: "both",
      animationIterationCount: 10,
      animationName: "slidein",
      animationPlayState: "running",
      animationTimeline: "test1",
      animationTimingFunction: "ease-in",
    });
  });
  test("should consume background props", () => {
    consumeAndTest({
      backgroundImage: "url('img_tree.png')",
      backgroundPosition: "center",
      backgroundPositionX: 10,
      backgroundPositionY: 20,
      backgroundSize: "cover",
      backgroundRepeat: "no-repeat",
      backgroundAttachment: "fixed",
      backgroundClip: "border-box",
      backgroundOrigin: "content-box",
      backgroundBlendMode: "hue",
    });
  });
  test("should consume shortened background props", () => {
    consumeAndTest({
      bgImage: "url('img_tree.png')",
      bgPosition: "center",
      bgPositionX: 10,
      bgPositionY: 20,
      bgSize: "cover",
      bgRepeat: "no-repeat",
      bgAttachment: "fixed",
      bgClip: "border-box",
      bgOrigin: "content-box",
      bgBlendMode: "hue",
    });
  });
  test("should consume border props", () => {
    consumeAndTest({
      borderX: 1,
      borderY: 2,
      border: "1px solid red",
      borderTop: 1,
      borderLeft: "2px",
      borderRight: 3,
      borderBottom: 4,
      borderRadius: 1,
      borderStyle: "solid",
      borderWidth: 10,
    });
  });
  test("should consume color props", () => {
    consumeAndTest({
      background: "red",
      backgroundColor: "red",
      color: "yellow",
      borderColor: "black",
      outlineColor: "pink",
      accentColor: "springgreen",
      caretColor: "salmon",
      opacity: 0.5,
      visibility: "hidden",
    });
  });
  test("should consume column props", () => {
    consumeAndTest({
      columnCount: 1,
      columnFill: "auto",
      columnGap: 1,
      columnRule: "1px solid red",
      columnRuleColor: "yellow",
      columnRuleStyle: "dotted",
      columnRuleWidth: 1,
      columnSpan: "all",
      columnWidth: 1,
      columns: "auto 2",
    });
  });
  test("should consume effect props", () => {
    consumeAndTest({
      transition: "0.5s ease",
      transitionDuration: "0.5s",
      transitionProperty: "all",
      transitionTimingFunction: "ease-in-out",
      transform: "scale(1.2)",
      transformBox: "border-box",
      transformOrigin: "center",
      transformStyle: "preserve-3d",
      clipPath: "ellipse(130px 140px at 10% 20%)",
    });
  });
  test("should consume filter props", () => {
    consumeAndTest({
      filter: "invert(100%)",
      backdropFilter: "blur(2px)",
    });
  });
  test("should consume flex props", () => {
    consumeAndTest({
      flexDirection: "column",
      justifyItems: "flex-start",
      justifyContent: "center",
      justifySelf: "flex-start",
      justify: "cener",
      alignContent: "center",
      alignItems: "center",
      alignSelf: "baseline",
      flex: "1 1 100px",
      flexBasis: 1,
      flexFlow: "row",
      flexGrow: 1,
      flexShrink: 1,
      flexWrap: "wrap",
      gap: "10px 20px",
      placeItems: "center",
      placeContent: "end center",
    });
  });
  test("should consume font props", () => {
    consumeAndTest({
      fontSize: 24,
      fontFamily: "Arial",
      fontWeight: 500,
    });
  });
  test("should consume grid props", () => {
    consumeAndTest({
      gridGap: 8,
      gridColumn: "2/3",
      gridColumnStart: 2,
      gridColumnEnd: 3,
      gridTemplateRows: "auto 1fr",
    });
  });
  test("should consume layout props", () => {
    consumeAndTest({
      width: 8,
      minWidth: "12px",
      height: "2em",
      maxHeight: "1rem",
      display: "flex",
      zIndex: 2,
      overflow: "hidden",
      overflowX: "auto",
      overflowY: "visible",
      aspectRatio: "16/9",
      boxSizing: "border-box",
      float: "right",
      clear: "both",
      objectFit: "contain",
      objectPosition: "250px 125px",
      resize: "horizontal",
      verticalAlign: "sub",
      userSelect: "none",
    });
    consumeAndTest({
      w: 8,
      minW: "12px",
      maxW: "2em",
      h: "2em",
      minH: "12px",
      maxH: "2rem",
    });
  });
  test("should consume list props", () => {
    consumeAndTest({
      listStyle: "outside",
      listStyleImage: "url('hoge.png')",
      listStylePosition: "inside",
      listStyleType: "disc",
    });
  });
  test("should consume mask props", () => {
    consumeAndTest({
      mask: "url(#c1) luminance",
      maskBorder: 'url("border-mask.png") 25',
      maskBorderMode: "alpha",
      maskBorderOutset: 1.5,
      maskBorderRepeat: "stretch round",
      maskBorderSlice: 10,
      maskBorderSource: "url('hogesrc.png')",
      maskBorderWidth: 10,
      maskClip: "border-box",
      maskComposite: "add",
      maskImage: "url('hogeimg.png')",
      maskMode: "luminance",
      maskOrigin: "border-box",
      maskPosition: "center",
      maskRepeat: "repeat",
      maskSize: "10% 20%",
      maskType: "alpha",
    });
  });
  test("should consume outline props", () => {
    consumeAndTest({
      outline: "1px solid red",
      outlineOffset: 1,
      outlineStyle: "dotted",
      outlineWidth: 1,
    });
  });
  test("should consume position props", () => {
    consumeAndTest({
      top: 1,
      right: "2px",
      left: "3rem",
      bottom: 4,
      inset: "10",
    });
  });

  test("should consume responsible styles", () => {
    consumeAndTest({
      fontSize: [24, 32],
    });
  });
  test("should consume styles with zero value", () => {
    consumeAndTest({
      fontSize: [0, 24],
      width: 0,
    });
  });
  test("should respect props order", () => {
    consumeAndTest({
      border: "1px solid",
      borderColor: "red",
    });
    consumeAndTest({
      borderColor: "red",
      border: "1px solid",
    });
  });
  test("should convert units", () => {
    consumeAndTest({
      width: 1,
    });
    consumeAndTest({
      width: "2px",
    });
    consumeAndTest({
      width: "3",
    });
    consumeAndTest({
      width: "4rem",
    });
  });
  test("should not include invalid keys in the resulting CSS", () => {
    consumeAndTest({
      invalid: true,
    } as unknown as StyledProps);
  });
});
