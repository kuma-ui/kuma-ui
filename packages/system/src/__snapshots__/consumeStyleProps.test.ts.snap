// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`consumeStyleProps function > should consume animation props 1`] = `
{
  "base": "animation: 3s ease-in 1s infinite reverse both running slidein;animation-composition: add;animation-delay: 1s;animation-direction: reverse;animation-duration: 1s;animation-fill-mode: both;animation-iteration-count: 10;animation-name: slidein;animation-play-state: running;animation-timeline: test1;animation-timing-function: ease-in;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume background props 1`] = `
{
  "base": "background-image: url('img_tree.png');background-position: center;background-position-x: 10px;background-position-y: 20px;background-size: cover;background-repeat: no-repeat;background-attachment: fixed;background-clip: border-box;background-origin: content-box;background-blend-mode: hue;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume border props 1`] = `
{
  "base": "border-left: 1px;border-right: 1px;border-top: 2px;border-bottom: 2px;border: 1px solid red;border-top: 1px;border-left: 2px;border-right: 3px;border-bottom: 4px;border-radius: 1px;border-style: solid;border-width: 10px;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume color props 1`] = `
{
  "base": "background: red;background-color: red;color: yellow;border-color: black;outline-color: pink;accent-color: springgreen;caret-color: salmon;opacity: 0.5;visibility: hidden;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume column props 1`] = `
{
  "base": "column-count: 1;column-fill: auto;column-gap: 1px;column-rule: 1px solid red;column-rule-color: yellow;column-rule-style: dotted;column-rule-width: 1px;column-span: all;column-width: 1px;columns: auto 2;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume effect props 1`] = `
{
  "base": "transition: 0.5s ease;transition-duration: 0.5s;transition-property: all;transition-timing-function: ease-in-out;transform: scale(1.2);transform-box: border-box;transform-origin: center;transform-style: preserve-3d;clip-path: ellipse(130px 140px at 10% 20%);",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume filter props 1`] = `
{
  "base": "filter: invert(100%);backdrop-filter: blur(2px);",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume flex props 1`] = `
{
  "base": "flex-direction: column;justify-items: flex-start;justify-content: center;justify-self: flex-start;justify-content: cener;align-content: center;align-items: center;align-self: baseline;flex: 1 1 100px;flex-basis: 1px;flex-flow: row;flex-grow: 1;flex-shrink: 1;flex-wrap: wrap;gap: 10px 20px;place-items: center;place-content: end center;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume font props 1`] = `
{
  "base": "font-size: 24px;font-family: Arial;font-weight: 500;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume grid props 1`] = `
{
  "base": "grid-gap: 8px;grid-column: 2/3;grid-column-start: 2;grid-column-end: 3;grid-template-rows: auto 1fr;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume layout props 1`] = `
{
  "base": "width: 8px;min-width: 12px;height: 2em;max-height: 1rem;display: flex;z-index: 2;overflow: hidden;overflow-x: auto;overflow-y: visible;aspect-ratio: 16/9;box-sizing: border-box;float: right;clear: both;object-fit: contain;object-position: 250px 125px;resize: horizontal;vertical-align: sub;user-select: none;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume layout props 2`] = `
{
  "base": "width: 8px;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume list props 1`] = `
{
  "base": "list-style: outside;list-style-image: url('hoge.png');list-style-position: inside;list-style-type: disc;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume mask props 1`] = `
{
  "base": "mask: url(#c1) luminance;mask-border: url(\\"border-mask.png\\") 25;mask-border-mode: alpha;mask-border-outset: 1.5px;mask-border-repeat: stretch round;mask-border-slice: 10px;mask-border-source: url('hogesrc.png');mask-border-width: 10px;mask-clip: border-box;mask-composite: add;mask-image: url('hogeimg.png');mask-mode: luminance;mask-origin: border-box;mask-position: center;mask-repeat: repeat;mask-size: 10% 20%;mask-type: alpha;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume outline props 1`] = `
{
  "base": "outline: 1px solid red;outline-offset: 1px;outline-style: dotted;outline-width: 1px;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume position props 1`] = `
{
  "base": "top: 1px;right: 2px;left: 3rem;bottom: 4px;inset: 10;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume responsible styles 1`] = `
{
  "base": "font-size: 24px;",
  "media": {
    "576px": "font-size: 32px;",
  },
}
`;

exports[`consumeStyleProps function > should consume shortened background props 1`] = `
{
  "base": "background-image: url('img_tree.png');background-position: center;background-position-x: 10px;background-position-y: 20px;background-size: cover;background-repeat: no-repeat;background-attachment: fixed;background-clip: border-box;background-origin: content-box;background-blend-mode: hue;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should consume styles with zero value 1`] = `
{
  "base": "font-size: 0px;width: 0px;",
  "media": {
    "576px": "font-size: 24px;",
  },
}
`;

exports[`consumeStyleProps function > should convert units 1`] = `
{
  "base": "width: 1px;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should convert units 2`] = `
{
  "base": "width: 2px;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should convert units 3`] = `
{
  "base": "width: 3;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should convert units 4`] = `
{
  "base": "width: 4rem;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should not include invalid keys in the resulting CSS 1`] = `
{
  "base": "",
  "media": {},
}
`;

exports[`consumeStyleProps function > should respect props order 1`] = `
{
  "base": "border: 1px solid;border-color: red;",
  "media": {},
}
`;

exports[`consumeStyleProps function > should respect props order 2`] = `
{
  "base": "border-color: red;border: 1px solid;",
  "media": {},
}
`;
